jwst/pipeline/calwebb_spec2.py

```diff
@@ -43,179 +48,226 @@ class Spec2Pipeline(Pipeline):
-
-            # Apply flat-field correction
-            input = self.flat_field(input)
-
-            # Apply the source type decision step
-            input = self.srctype(input)
-
-            # Apply the straylight correction for MIRI MRS
-            if exp_type == 'MIR_MRS':
-                input = self.straylight(input)
-
-            # Apply the fringe correction for MIRI MRS
-            if exp_type == 'MIR_MRS':
-                input = self.fringe(input)
-
-            # Apply pathloss correction to NIRSpec exposures
-            if exp_type in ['NRS_FIXEDSLIT', 'NRS_BRIGHTOBJ', 'NRS_MSASPEC',
-                            'NRS_IFU']:
-                input = self.pathloss(input)
-
-            # Apply flux calibration
-            input = self.photom(input)
-
-            # Record ASN pool and table names in output
-            if input_table.poolname:
-                input.meta.asn.pool_name = input_table.poolname
-                input.meta.asn.table_name = input_table.filename
-            else:
-                input.meta.asn.pool_name = ' '
-                input.meta.asn.table_name = ' '
+        asn = Level2Association.open(input)
+
+        # Each exposure is a product in the association.
+        # Process each exposure.
+        for product in asn['products']:
+            log.info('Processing exposure {}'.format(product['name']))
+            self.process_exposure_product(
+                product,
+                asn['asn_pool'],
+                asn.filename
+            )
 
-            # Save the calibrated exposure
-            if isinstance(input, datamodels.CubeModel):
-                self.save_model(input, 'calints')
-            else:
-                self.save_model(input, "cal")
-            log.info('Saved calibrated product to %s' % input.meta.filename)
+        # We're done
+        log.info('Ending calwebb_spec2')
+
+    # Process each exposure
+    def process_exposure_product(
+            self,
+            exp_product,
+            pool_name=' ',
+            asn_file=' '
+    ):
+        """Process an exposure found in the association product
+
+        Parameters
+        ---------
+        exp_product: dict
+            A Level2b association product.
+        """
+
+        # Find all the member types in the product
+        members_by_type = defaultdict(list)
+        for member in exp_product['members']:
+            members_by_type[member['exptype']].append(member['expname'])
+
+        # Get the science member. Technically there should only be
+        # one. We'll just get the first one found.
+        science = members_by_type['SCIENCE']
+        if len(science) != 1:
+            log.warn(
+                'Wrong number of science files found in {}'.format(
+                    exp_product['name']
+                )
+            )
+        science = science[0]
+
+        self.log.info('Working on input %s ...', science)
+        if isinstance(science, datamodels.DataModel):
```

Comment:
Is this if-block necessary? I believe this functionality is already built into datamodels.open, such that it tests to see if the input is a string or data model. If it's a string, it loads it into a data model, else if it's already a data model, it simply returns without doing anything.